### PR TITLE
[DC21X4] Add driver for DC21X4 adapters

### PR DESCRIPTION
**Pull request by disean**
## Purpose

This patch adds a driver for the DC21x4 compatible (these are really rare nowadays) adapters, that were common in DEC Alpha boxes. The 21140 chip is emulated in VPC 2007/Hyper-V. This is an experimental driver, not yet tested on real hardware.

JIRA issue: [CORE-8724](https://jira.reactos.org/browse/CORE-8724)

### CODE REVIEW

**Comment by binarymaster** [comment]
Some build problems:

```
/home/runner/work/reactos/reactos/RosBE-CI/i386/lib/gcc/i686-w64-mingw32/8.4.0/../../../../i686-w64-mingw32/bin/ld: drivers/network/dd/dc21x4/CMakeFiles/dc21x4.dir/eeprom.c.obj:eeprom.c:(PAGE+0x227): undefined reference to `MediaNumber2Str'
/home/runner/work/reactos/reactos/RosBE-CI/i386/lib/gcc/i686-w64-mingw32/8.4.0/../../../../i686-w64-mingw32/bin/ld: drivers/network/dd/dc21x4/CMakeFiles/dc21x4.dir/eeprom.c.obj:eeprom.c:(PAGE+0x75e): undefined reference to `MediaNumber2Str'
/home/runner/work/reactos/reactos/RosBE-CI/i386/lib/gcc/i686-w64-mingw32/8.4.0/../../../../i686-w64-mingw32/bin/ld: drivers/network/dd/dc21x4/CMakeFiles/dc21x4.dir/eeprom.c.obj:eeprom.c:(PAGE+0x915): undefined reference to `MediaNumber2Str'
/home/runner/work/reactos/reactos/RosBE-CI/i386/lib/gcc/i686-w64-mingw32/8.4.0/../../../../i686-w64-mingw32/bin/ld: drivers/network/dd/dc21x4/CMakeFiles/dc21x4.dir/media.c.obj:media.c:(.text+0x1c): undefined reference to `MediaNumber2Str'
/home/runner/work/reactos/reactos/RosBE-CI/i386/lib/gcc/i686-w64-mingw32/8.4.0/../../../../i686-w64-mingw32/bin/ld: drivers/network/dd/dc21x4/CMakeFiles/dc21x4.dir/media.c.obj:media.c:(.text+0x1ca): undefined reference to `MediaNumber2Str'
/home/runner/work/reactos/reactos/RosBE-CI/i386/lib/gcc/i686-w64-mingw32/8.4.0/../../../../i686-w64-mingw32/bin/ld: drivers/network/dd/dc21x4/CMakeFiles/dc21x4.dir/media.c.obj:media.c:(.text+0x309): more undefined references to `MediaNumber2Str' follow
```

File: drivers/network/dd/dc21x4/dc21x4.h [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+
+#if defined(MEMCMP_HACK)
+BOOLEAN
+DcEqualMemory(
+    _In_reads_bytes_(Length) const VOID* Destination,
+    _In_reads_bytes_(Length) const VOID* Source,
+    _In_ SIZE_T Length);
+#undef NdisEqualMemory
+#define NdisEqualMemory DcEqualMemory
+#endif
```
Comment (binarymaster) [inline drivers/network/dd/dc21x4/dc21x4.h:482]: Why this hack is needed, does `NdisEqualMemory` function has some problem in GCC builds? (and I've just noticed a similar workaround in nvnet)

File: drivers/network/dd/dc21x4/init.c [hunk 2]
```diff
@@ -0,0 +1,1147 @@
...
+                                       sizeof(Buffer));
+    if (Bytes != sizeof(Buffer))
+        return NDIS_STATUS_FAILURE;
+
+    Adapter->DeviceId = PciConfig->DeviceID;
+    Adapter->RevisionId = PciConfig->RevisionID;
+
+    switch ((PciConfig->DeviceID << 16) | PciConfig->VendorID)
+    {
+        case 0x00021011: /* 21040 */
```
Comment (binarymaster) [inline drivers/network/dd/dc21x4/init.c:768]: You can add constants in headers for known PCI device identifiers.

File: drivers/network/dd/dc21x4/dc21x4.c [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    ULONG i, j, Crc;
+    const UCHAR* Data = Buffer;
+
+    Crc = 0xFFFFFFFF;
+    for (i = 0; i < Size; ++i)
+    {
+        Crc ^= Data[i];
+        for (j = 8; j > 0; j--)
+        {
+            Crc = (Crc >> 1) ^ (-(LONG)(Crc & 1) & 0xEDB88320);
```
[the remark this answers is not in the document]
Comment (ThFabba) [reply drivers/network/dd/dc21x4/dc21x4.c:51]: Why?

[the remark this answers is not in the document]
Comment (binarymaster) [reply drivers/network/dd/dc21x4/dc21x4.c:51]: A documented constant is always better, even if used in one place I think.

File: drivers/network/dd/dc21x4/dc21x4.h [hunk 4]
```diff
@@ -0,0 +1,536 @@
+/*
+ * PROJECT:     ReactOS DC21x4 Driver
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Main header file
+ * COPYRIGHT:   Copyright 2023 Dmitry Borisov <di.sean@protonmail.com>
+ */
+
+#ifndef _DC21X4_PCH_
+#define _DC21X4_PCH_
```
Comment (JoachimHenze) [inline drivers/network/dd/dc21x4/dc21x4.h:9]: you used pragma once in most of your headers. Possible also here? Or is there any specific reasons why classic include-guards are the only option here?

File: drivers/network/dd/dc21x4/dc21x4.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.c:52]: Added a comment.

File: drivers/network/dd/dc21x4/init.c (hunk 2 above)
Comment (disean) [reply drivers/network/dd/dc21x4/init.c]: Done.

File: drivers/network/dd/dc21x4/power.c [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    _In_ PDC21X4_ADAPTER Adapter)
+{
+    ULONG WakeUpControl;
+
+    PAGED_CODE();
+
+    /* Clear the wake-up events */
+    WakeUpControl = (DC_WAKE_UP_STATUS_LINK_CHANGE |
+                     DC_WAKE_UP_STATUS_MAGIC_PACKET |
+                     DC_WAKE_UP_STATUS_LINK_CHANGE);
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/power.c]: Fixed, thanks.

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 1 above)
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.h]: Without this hack, it will not compile in debug mode.

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 4 above)
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.h:9]: Is it allowed to use #pragma once in a precompiled header file?

**Comment by binarymaster** [comment]
from `build-linux (clang, Debug)`:

```
../src/drivers/network/dd/dc21x4/dc21x4.c:230:9: error: use of undeclared identifier '__RELFILE__'
        ERR("Hardware is gone...\n");
        ^
../src/drivers/network/dd/dc21x4/debug.h:65:39: note: expanded from macro 'ERR'
      if (DbgPrint("(%s:%d) %s " fmt, __RELFILE__, __LINE__, __FUNCTION__, ##__VA_ARGS__)) \
                                      ^
```

from `build-msvc (windows-2019, 14.0)`:

```
D:\a\reactos\reactos\src\drivers\network\dd\dc21x4\dc21x4.c(191): error C2983: 'DcStartAdapter': all declarations must have an identical __declspec(code_seg(...))
D:/a/reactos/reactos/src/drivers/network/dd/dc21x4/dc21x4.h(396): note: see declaration of 'DcStartAdapter'
```

File: drivers/network/dd/dc21x4/dc21x4.h [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+    _In_opt_ PVOID Context);
+
+DECLSPEC_NOINLINE
+VOID
+DcStopAdapter(
+    _In_ PDC21X4_ADAPTER Adapter,
+    _In_ BOOLEAN WaitForPackets);
+
+VOID
+DcStartAdapter(
```
Comment (binarymaster) [inline drivers/network/dd/dc21x4/dc21x4.h:396]:
```suggestion
CODE_SEG("PAGE")
VOID
DcStartAdapter(
```

**Review by binarymaster — APPROVED** [review]

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 4 above)
Comment (HBelusca) [reply drivers/network/dd/dc21x4/dc21x4.h:9]: iirc. yes it's possible.

**Review by GeoB99 — APPROVED** [review]

**Comment by HBelusca** [comment]
I'll try to review this PR tomorrow.

**Review by JoachimHenze — APPROVED** [review]

File: drivers/network/dd/dc21x4/eeprom.c [hunk 7]
```diff
@@ -0,0 +1,1527 @@
...
+    ULONG Data, i, j;
+
+    PAGED_CODE();
+
+    /* Reset the ROM pointer */
+    DC_WRITE(Adapter, DcCsr9_SerialInterface, 0);
+
+    for (i = 0; i < EAR_SIZE; ++i)
+    {
+        for (j = 10000; j > 0; --j)
```
Comment (HBelusca) [inline drivers/network/dd/dc21x4/eeprom.c:1364]: Why 10000?

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 1 above)
Comment (HBelusca) [reply drivers/network/dd/dc21x4/dc21x4.h]: What's the error you obtain when you compile in debug mode?

Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.h]:
```
undefined reference to `memcmp'
collect2.exe: error: ld returned 1 exit status
```

File: drivers/network/dd/dc21x4/media140.c [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        if ((Adapter->MediaBitmap & (1 << i)) && !MEDIA_IS_FD(i))
+        {
+            return i;
+        }
+    }
+}
+
+VOID
+NTAPI
+MediaMonitor21140Dpc(
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/media140.c:82]: The timer callback is always called at DPC level.

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 4 above)
Comment (JoachimHenze) [reply drivers/network/dd/dc21x4/dc21x4.h:9]: Then let's do it.

File: drivers/network/dd/dc21x4/eeprom.c [hunk 9]
```diff
@@ -0,0 +1,1527 @@
...
+{
+    PLIST_ENTRY ListHead, NextEntry;
+    PDC_SROM_ENTRY SRomEntry;
+
+    PAGED_CODE();
+
+    /* Loop the adapter list backwards */
+    ListHead = &SRompAdapterList;
+    NextEntry = ListHead->Blink;
+    while (NextEntry != ListHead)
```
[the remark this answers is not in the document]
Comment (JoachimHenze) [reply drivers/network/dd/dc21x4/eeprom.c:76]: Would the for-loop-conditions all fit into a single line if we named it just "prev" (or "next" whatever is more correct).
I find this way of splitting the condiitions into several lines very unfriendly to read.

[the remark this answers is not in the document]
Comment (HBelusca) [reply drivers/network/dd/dc21x4/eeprom.c:76]: on the contrary, this avoid having super-long lines, and each condition is in its own line, so it's very clear what's going on...

File: drivers/network/dd/dc21x4/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,37 @@
+
+add_definitions(
+    -DNDIS_MINIPORT_DRIVER
+    -DNDIS51_MINIPORT=1)
+
+list(APPEND SOURCE
+    dc21x4.c
+    dc21x4.h
+    dc21x4hw.h
```
[the remark this answers is not in the document]
Comment (JoachimHenze) [reply drivers/network/dd/dc21x4/CMakeLists.txt:9]: Doesn't that point towards some of these includes missing in some of the .c files then? Or how can that be explained?

[the remark this answers is not in the document]
Comment (HBelusca) [reply drivers/network/dd/dc21x4/CMakeLists.txt:9]: What do you mean? Not all .c files there need what's exposed by all those headers.

File: drivers/network/dd/dc21x4/eeprom.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (JoachimHenze) [reply drivers/network/dd/dc21x4/eeprom.c:76]: At first glance the condition looks like a command. You have to look twice to see it being part of the conditions.

File: drivers/network/dd/dc21x4/eeprom.c [hunk 11]
```diff
@@ -0,0 +1,1527 @@
...
+Exit:
+    SRomReleaseListMutex();
+
+    return Found;
+}
+
+static
+CODE_SEG("PAGE")
+BOOLEAN
+SRomIsEmpty(
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/eeprom.c:205]: Yes, but on multiport boards the EAR data usually contains 0xFF in all locations.

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 1 above)
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.h]: Added a new library to the list.

File: drivers/network/dd/dc21x4/send.c [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+
+    Destination = Buffer->VirtualAddress;
+
+    while (TRUE)
+    {
+        NdisMoveMemory(Destination, Address, CurrentLength);
+        Destination += CurrentLength;
+
+        NdisGetNextBuffer(CurrentBuffer, &CurrentBuffer);
+        if (!CurrentBuffer)
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/send.c:111]: It's up to ndis.sys/protocol drivers to make sure the MDL chain is "safe".

File: drivers/network/dd/dc21x4/dc21x4hw.h [hunk 13]
```diff
@@ -0,0 +1,559 @@
...
+
+#include <poppack.h>
+
+/*
+ * NIC Control and Status Registers
+ */
+typedef enum _DC_CSR
+{
+    DcCsr0_BusMode = 0x00,
+#define DC_BUS_MODE_SOFT_RESET                 0x00000001
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4hw.h]: Done. I feel like that's not a real solution.

File: drivers/network/dd/dc21x4/init.c [hunk 14]
```diff
@@ -0,0 +1,1319 @@
...
+                              Rcb->VirtualAddressOriginal,
+                              Rcb->PhysicalAddressOriginal);
+    }
+
+    if (Rcb->NdisBuffer)
+        NdisFreeBuffer(Rcb->NdisBuffer);
+    if (Rcb->Packet)
+        NdisFreePacket(Rcb->Packet);
+
+    NdisFreeMemory(Rcb, sizeof(*Rcb), 0);
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/init.c:174]: NdisFreeMemoryWithTag is an NDIS 6.2+ API.

File: drivers/network/dd/dc21x4/dc21x4.h (hunk 4 above)
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.h]: Done.

File: drivers/network/dd/dc21x4/eeprom.h [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    UCHAR BlockCount;
+} DC_SROM_COMPACT_BLOCK, *PDC_SROM_COMPACT_BLOCK;
+#include <poppack.h>
+
+typedef struct _DC_SROM_REPAIR_ENTRY
+{
+    PUCHAR InfoLeaf;
+#if DBG
+    PCSTR Name;
+#endif
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/eeprom.h:23]: The size of PCSTR is 8 bytes on a 64-bit system.

File: drivers/network/dd/dc21x4/send.c [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+
+    for (Entry = Adapter->SendQueueList.Flink;
+         Entry != &Adapter->SendQueueList;
+         Entry = Entry->Flink)
+    {
+        PNDIS_PACKET Packet = DC_PACKET_FROM_LIST_ENTRY(Entry);
+
+        if (NDIS_GET_PACKET_CANCEL_ID(Packet) == CancelId)
+        {
+            RemoveEntryList(DC_LIST_ENTRY_FROM_PACKET(Packet));
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/send.c:297]: Good catch!

File: drivers/network/dd/dc21x4/util.h [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+{
+    UCHAR Destination[ETH_LENGTH_OF_ADDRESS];
+    UCHAR Source[ETH_LENGTH_OF_ADDRESS];
+    USHORT PayloadType;
+} ETH_HEADER, *PETH_HEADER;
+
+typedef struct _MY_USHORT
+{
+    USHORT Word;
+} MY_USHORT, *PMY_USHORT;
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/util.h]: From what I've heard, GCC (?) has support for unaligned access through packed structures. Removed.

File: drivers/network/dd/dc21x4/phy.c [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+        INFO_VERB("Found PHY at address %u: ID %04lx:%04lx, Ctrl %04lx, Status %04lx, Adv %04lx\n",
+                  PhyAddress,
+                  PhyIdLow,
+                  PhyIdHigh,
+                  MiiControl,
+                  MiiStatus,
+                  MiiAdvertise);
+
+        /* Sanity check */
+        ASSERT((MiiStatus & 0xFF00) == 0x7800);
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/phy.c]: Removed. QEMU doesn't report the correct MII status.

File: drivers/network/dd/dc21x4/hardware.c [hunk 19]
```diff
@@ -0,0 +1,585 @@
...
+DcStopTxRxProcess(
+    _In_ PDC21X4_ADAPTER Adapter)
+{
+    ULONG i, OpMode, Status;
+
+    OpMode = Adapter->OpMode;
+    OpMode &= ~(DC_OPMODE_RX_ENABLE | DC_OPMODE_TX_ENABLE);
+    DC_WRITE(Adapter, DcCsr6_OpMode, OpMode);
+
+    for (i = 0; i < 10000; ++i)
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/hardware.c]: It's not clear how much time is needed to wait for a result. We cannot wait forever though.

File: drivers/network/dd/dc21x4/send.c [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+
+VOID
+DcProcessPendingPackets(
+    _In_ PDC21X4_ADAPTER Adapter)
+{
+    PLIST_ENTRY Entry;
+    NDIS_STATUS Status;
+    PNDIS_PACKET Packet;
+
+    do
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/send.c:210]: Fixed

File: drivers/network/dd/dc21x4/util.h [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+#else
+    USHORT Result;
+
+    NdisMoveMemory(&Result, Data, sizeof(Result));
+    return Result;
+#endif
+}
+
+/*
+ * This is an optimized version of the PopEntryList() function.
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/util.h:65]: I wouldn't expect a noticeable difference between them. Added a DBG switch.

File: drivers/network/dd/dc21x4/dc21x4.c [hunk 22]
```diff
@@ -0,0 +1,422 @@
...
+#if DBG
+    NdisAcquireSpinLock(&Adapter->ReceiveLock);
+    if (Adapter->RcbFree != Adapter->RcbCount)
+    {
+        INFO("RX packets: %u/%u\n", Adapter->RcbFree, Adapter->RcbCount);
+    }
+    NdisReleaseSpinLock(&Adapter->ReceiveLock);
+#endif
+
+    while (TRUE)
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.c:106]: Done

File: drivers/network/dd/dc21x4/init.c [hunk 23]
```diff
@@ -0,0 +1,1319 @@
...
+    Rcb->PhysicalAddress = ALIGN_UP_BY(PhysicalAddress.LowPart, SYSTEM_CACHE_ALIGNMENT_SIZE);
+
+    ASSERT((Rcb->PhysicalAddress % DC_RECEIVE_BUFFER_ALIGNMENT) == 0);
+
+    Rcb->VirtualAddressOriginal = VirtualAddress;
+    Rcb->PhysicalAddressOriginal.QuadPart = PhysicalAddress.QuadPart;
+
+    NdisAllocatePacket(&Status, &Rcb->Packet, Adapter->PacketPool);
+    if (Status != NDIS_STATUS_SUCCESS)
+        goto Failure;
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/init.c:216]: It wouldn't leak in this situation.

File: drivers/network/dd/dc21x4/dc21x4.c [hunk 24]
```diff
@@ -0,0 +1,422 @@
...
+         Tcb = DC_NEXT_TCB(Adapter, Tcb))
+    {
+        Packet = Tcb->Packet;
+
+        if (!Packet)
+            continue;
+
+        InsertTailList(&DoneList, DC_LIST_ENTRY_FROM_PACKET(Packet));
+
+        DC_RELEASE_TCB(Adapter, Tcb);
```
[the remark this answers is not in the document]
Comment (disean) [reply drivers/network/dd/dc21x4/dc21x4.c:64]: Yes

File: drivers/network/dd/dc21x4/CMakeLists.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (binarymaster) [reply drivers/network/dd/dc21x4/CMakeLists.txt:9]: > which .c is not rebuilt then specifically?

I just remember there were such problems from time to time, but currently I can't recall in what modules.

File: drivers/network/dd/dc21x4/hardware.c [hunk 25]
```diff
@@ -0,0 +1,585 @@
...
+    Tbd->Control |= Control;
+    DC_WRITE_BARRIER();
+    Tbd->Status = DC_TBD_STATUS_OWNED;
+
+    DC_WRITE(Adapter, DcCsr1_TxPoll, DC_TX_POLL_DOORBELL);
+
+    if (!WaitForCompletion)
+        return TRUE;
+
+    for (i = 100000; i > 0; --i)
```
[the remark this answers is not in the document]
Comment (binarymaster) [reply drivers/network/dd/dc21x4/hardware.c:240]: Could be improved by mentioning units (possibly adding a constant for seconds).

File: drivers/network/dd/dc21x4/net21x4.inf [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+
+NA="Network Address"
+SD="Speed & Duplex"
+Auto="Auto"
+
+DC21040.DeviceDesc="Intel 21040 PCI Ethernet Adapter"
+DC21041.DeviceDesc="Intel 21041 PCI Ethernet Adapter"
+DC21140.DeviceDesc="Intel 21140 PCI Ethernet Adapter"
+DC21143.DeviceDesc="Intel 21143 PCI Ethernet Adapter"
+DC21145.DeviceDesc="Intel 21145 PCI Ethernet Adapter"
```
Comment (binarymaster) [inline drivers/network/dd/dc21x4/net21x4.inf:144]: Were they all manufactured by Intel? Maybe some of them actually by DECchip? Also I remember Windows 7 has a distinct entry (perhaps based on subsystem value) detecting an emulated adapter by Virtual PC.

Comment (binarymaster) [reply drivers/network/dd/dc21x4/net21x4.inf:144]: Found it: `PCI\VEN_1011&DEV_0009&SUBSYS_21140A00&REV_20` - detection string for the emulated adapter. (Source: https://community.osr.com/discussion/comment/242520/#Comment_242520)

Comment (disean) [reply drivers/network/dd/dc21x4/net21x4.inf]: Intel bought DEC in 1997 :)

**Pull request merged**